src/ctap/credential_management.rs

```diff
@@ -0,0 +1,900 @@
+    if total_rps > 1 {
+        *stateful_command_permission =
+            TimedPermission::granted(now, STATEFUL_COMMAND_TIMEOUT_DURATION);
+        *stateful_command_type = Some(StatefulCommand::EnumerateRps(rp_ids));
+    }
+    enumerate_rps_response(rp_id, Some(total_rps as u64))
+}
+
+fn process_enumerate_rps_get_next_rp(
+    stateful_command_permission: &mut TimedPermission,
+    stateful_command_type: &mut Option<StatefulCommand>,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    check_command_permission(stateful_command_permission, now)?;
+    let rp_id = if let Some(StatefulCommand::EnumerateRps(rp_ids)) = stateful_command_type {
+        rp_ids.pop().ok_or(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED)?
+    } else {
+        return Err(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED);
+    };
+    enumerate_rps_response(Some(rp_id), None)
+}
+
+fn process_enumerate_credentials_begin(
+    persistent_store: &PersistentStore,
+    stateful_command_permission: &mut TimedPermission,
+    stateful_command_type: &mut Option<StatefulCommand>,
+    sub_command_params: CredentialManagementSubCommandParameters,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    let rp_id_hash = sub_command_params
+        .rp_id_hash
+        .ok_or(Ctap2StatusCode::CTAP2_ERR_NO_CREDENTIALS)?;
+    // TODO(kaczmarczyck) filter by rp_id_hash inside of storage?
+    let mut rp_credentials: Vec<PublicKeyCredentialSource> = persistent_store
+        .all_credentials()?
+        .into_iter()
+        .filter(|cred| {
+            let cred_rp_id_hash = Sha256::hash(cred.rp_id.as_bytes());
+            rp_id_hash == cred_rp_id_hash
+        })
+        .collect();
+    let total_credentials = rp_credentials.len();
+
+    let credential = rp_credentials
+        .pop()
+        .ok_or(Ctap2StatusCode::CTAP2_ERR_NO_CREDENTIALS)?;
+    if total_credentials > 1 {
+        *stateful_command_permission =
+            TimedPermission::granted(now, STATEFUL_COMMAND_TIMEOUT_DURATION);
+        *stateful_command_type = Some(StatefulCommand::EnumerateCredentials(rp_credentials));
+    }
+    enumerate_credentials_response(credential, Some(total_credentials as u64))
+}
+
+fn process_enumerate_credentials_get_next_credential(
+    stateful_command_permission: &mut TimedPermission,
+    mut stateful_command_type: &mut Option<StatefulCommand>,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    check_command_permission(stateful_command_permission, now)?;
+    let credential = if let Some(StatefulCommand::EnumerateCredentials(rp_credentials)) =
+        &mut stateful_command_type
+    {
+        rp_credentials
+            .pop()
+            .ok_or(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED)?
+    } else {
+        return Err(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED);
+    };
+    enumerate_credentials_response(credential, None)
```

Comment:
Same as before, I'm surprised we return `None` even if `rp_credentials` might be longer than 1.

Reply:
Same as above, I'll move the function call inside the `if`. The `None` is intended.

See [the spec](https://fidoalliance.org/specs/fido-v2.1-rd-20201208/fido-client-to-authenticator-protocol-v2.1-rd-20201208.html#authenticatorCredentialManagement), 6.8.3 and 6.8.4.

Reply:
Yeah, I confused this parameter with the remaining items to return. All good.